pump/storage/storage.go

```diff
@@ -1205,30 +1232,40 @@ func (a *Append) PullCommitBinlog(ctx context.Context, last int64) <-chan []byte
 						}
 
 						errorCount.WithLabelValues("feed_pre_write_value").Add(1.0)
-						log.Error("feed pre write value failed", zap.Error(err))
 						iter.Release()
+						a.gcTS.ReleaseLoadLock()
+						log.Error("feed pre write value failed", zap.Int64("TS", decodeTSKey(iter.Key())), zap.Error(err))
+						errs <- errors.Errorf("feed pre write value failed, TS %d", decodeTSKey(iter.Key()))
 						return
 					}
 				}
 
 				value, err = binlog.Marshal()
 				if err != nil {
-					log.Error("marshal failed", zap.Error(err))
 					iter.Release()
+					a.gcTS.ReleaseLoadLock()
+					log.Error("marshal failed", zap.Int64("TS", decodeTSKey(iter.Key())), zap.Error(err))
+					errs <- errors.Errorf("marshal failed, TS %d", decodeTSKey(iter.Key()))
 					return
 				}
 
 				select {
 				case values <- value:
 					log.Debug("send value success")
+				case <-time.After(gcMaxBlockTime):
+					// do not update `last` anymore.
+					log.Warn("can not send the binlog for a long time, will try to read again", zap.Duration("duration", gcMaxBlockTime), zap.Int64("current TS", decodeTSKey(iter.Key())))
```

Comment:
should we send error to `errs` here?

Reply:
`send the binlog for a long time` often caused by a slow drainer. if we send errors, the replication will be broken and try to send again, that's should make it worse.

Reply:
so, if drainer encounter the purge [error](https://github.com/pingcap/tidb-binlog/pull/1017/files#diff-23cbf283b94e3be793debbeba5eb816038ee3db75881faf0130e2a8e165cd6e8R141) ，all pump will reach this warning 🤔️, but both drainer and pump still running. until the user manually fix it

Reply:
? not **all** pump, but only this one. and pump should still be running because may other drainers can still work normally.

Reply:
if one pump is blocked, it seems the merge source will [blocked](https://github.com/pingcap/tidb-binlog/blob/0199ccdcd301ba5c7e86e67c78c1462372c9442b/drainer/merge.go#L277) soon.

Reply:
oh, I know what you mean. any advice?